apps/backend/src/libs/hooks/debouncer/set-rate-limit.hook.ts

```diff
@@ -0,0 +1,41 @@
+import { type UserAuthResponseDto } from "~/libs/types/types.js";
+import { AuthError } from "~/modules/auth/libs/exceptions/unauthorized.exception.js";
+
+import { type APIPreHandler } from "../libs/types/types.js";
+import { DEBOUNCER_CONFIG } from "./libs/constants/constants.js";
+import { RequestLimitError } from "./libs/exceptions/exceptions.js";
+
+const INCREMENT = 1;
+
+const setRateLimit = (requestsLimitPerMinute: number): APIPreHandler => {
+	const userIdToRequestNumber: Map<UserAuthResponseDto["id"], number> =
+		new Map();
+
+	return ({ user }, _, done): void => {
+		if (!user) {
+			throw new AuthError();
+		}
+
+		const { id } = user;
+		const requestsAmount = userIdToRequestNumber.get(id);
+
+		if (!requestsAmount) {
+			userIdToRequestNumber.set(id, INCREMENT);
+
+			setTimeout(() => {
+				userIdToRequestNumber.set(id, 0);
+			}, DEBOUNCER_CONFIG.RESET_TIME);
+
+			return;
+		}
+
+		if (requestsAmount === requestsLimitPerMinute) {
```

Comment:
it's always better to account for non standard concurrency cases
```suggestion
		if (requestsAmount >= requestsLimitPerMinute) {
```